File: iOS/Summit/Summit/Http/Requester/UserRequester.swift
//
//  UserRequester.swift
//  Summit
//
//  Created by Leapfrog-Software on 2018/04/29.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

enum AgeType {
    case u20
    case s20
    case s30
    case s40
    case s50
    case o60
    
    static func create(value: String) -> AgeType {
        switch value {
        case "0":
            return .u20
        case "1":
            return .s20
        case "2":
            return .s30
        case "3":
            return .s40
        case "4":
            return .s50
        default:
            return .o60
        }
    }
    
    func toString() -> String {
        switch self {
        case .u20:
            return "0"
        case .s20:
            return "1"
        case .s30:
            return "2"
        case .s40:
            return "3"
        case .s50:
            return "4"
        default:
            return "5"
        }
    }
}

enum GenderType {
    case male
    case female
    case unknown
    
    static func create(value: String) -> GenderType {
        switch value {
        case "0":
            return .male
        case "1":
            return .female
        default:
            return .unknown
        }
    }
    
    func toString() -> String {
        switch self {
        case .male:
            return "0"
        case .female:
            return "1"
        default:
            return "2"
        }
    }
}

struct UserData {
    
    let userId: String
    var nameFirst: String
    var nameLast: String
    var kanaFirst: String
    var kanaLast: String
    var age: AgeType
    var gender: GenderType
    var job: String
    var company: String
    var position: String
    var reserves: [String]
    var cards: [String]
    
    init?(data: Dictionary<String, Any>) {
        
        guard let userId = data["id"] as? String else {
            return nil
        }
        self.userId = userId
        
        self.nameFirst = data["nameFirst"] as? String ?? ""
        self.nameLast = data["nameLast"] as? String ?? ""
        self.kanaFirst = data["kanaFirst"] as? String ?? ""
        self.kanaLast = data["kanaLast"] as? String ?? ""
        self.age = AgeType.create(value: (data["age"] as? String ?? ""))
        self.gender = GenderType.create(value: (data["gender"] as? String ?? ""))
        self.company = data["company"] as? String ?? ""
        self.job = data["job"] as? String ?? ""
        self.position = data["position"] as? String ?? ""
        self.reserves = data["reserves"] as? Array<String> ?? []
        self.cards = data["cards"] as? Array<String> ?? []
    }
}

class UserRequester {
    
    static let shared = UserRequester()
    
    var dataList = [UserData]()
    
    func fetch(completion: @escaping ((Bool) -> ())) {
        
        let params = ["command": "getUser"]
        ApiManager.post(params: params) { [weak self] result, data in
            if result, let users = data as? Array<Dictionary<String, Any>> {
                self?.dataList = users.flatMap { UserData(data: $0) }
                completion(true)
            } else {
                completion(false)
            }
        }
    }
    
    func myUserData() -> UserData? {
        return self.dataList.filter { $0.userId == SaveData.shared.userId }.first
    }
    
    func query(userId: String) -> UserData? {
        return self.dataList.filter { $0.userId == userId }.first
    }
}

File: Android/app/src/main/java/leapfrog_inc/summit/Http/Requester/UserRequester.java
package leapfrog_inc.summit.Http.Requester;

import android.util.Base64;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import leapfrog_inc.summit.Function.Base64Utility;
import leapfrog_inc.summit.Function.Constants;
import leapfrog_inc.summit.Function.KanaUtils;
import leapfrog_inc.summit.Function.SaveData;
import leapfrog_inc.summit.Http.HttpManager;
import leapfrog_inc.summit.Http.Requester.Enum.AgeType;
import leapfrog_inc.summit.Http.Requester.Enum.GenderType;

/**
 * Created by Leapfrog-Software on 2018/05/08.
 */

public class UserRequester {

    public static class UserData {

        public String userId = "";
        public String nameFirst = "";
        public String nameLast = "";
        public String kanaFirst = "";
        public String kanaLast = "";
        public AgeType age = AgeType.unknown;
        public GenderType gender = GenderType.unknown;
        public String job = "";
        public String company = "";
        public String position = "";
        public ArrayList<String> reserves = new ArrayList<String>();
        public ArrayList<String> cards = new ArrayList<String>();
        public String message = "";

        static public UserData create(JSONObject json) {

            try {
                String userId = json.getString("id");
                String nameFirst = Base64Utility.decode(json.getString("nameFirst"));
                String nameLast = Base64Utility.decode(json.getString("nameLast"));
                String kanaFirst = Base64Utility.decode(json.getString("kanaFirst"));
                String kanaLast = Base64Utility.decode(json.getString("kanaLast"));
                AgeType age = AgeType.create(json.getString("age"));
                GenderType gender = GenderType.create(json.getString("gender"));
                String job = Base64Utility.decode(json.getString("job"));
                String company = Base64Utility.decode(json.getString("company"));
                String position = Base64Utility.decode(json.getString("position"));
                JSONArray reserves = json.getJSONArray("reserves");
                JSONArray cards = json.getJSONArray("cards");
                String message = Base64Utility.decode(json.getString("message"));

                UserData userData = new UserData();
                userData.userId = userId;
                userData.nameFirst = nameFirst;
                userData.nameLast = nameLast;
                userData.kanaFirst = kanaFirst;
                userData.kanaLast = kanaLast;
                userData.age = age;
                userData.gender = gender;
                userData.job = job;
                userData.company = company;
                userData.position = position;

                for (int i = 0; i < reserves.length(); i++) {
                    String reserve = reserves.getString(i);
                    if (!userData.reserves.contains(reserve)) {
                        userData.reserves.add(reserve);
                    }
                }

                for (int i = 0; i < cards.length(); i++) {
                    String card = cards.getString(i);
                    if (!userData.cards.contains(card)) {
                        userData.cards.add(card);
                    }
                }

                userData.message = message;

                return userData;

            } catch(Exception e) {}

            return null;
        }
    }

    private static UserRequester requester = new UserRequester();

    private UserRequester(){}

    public static UserRequester getInstance(){
        return requester;
    }

    private ArrayList<UserData> mDataList = new ArrayList<UserData>();

    public void fetch(final UserRequesterCallback callback){

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        if (ret.equals("0")) {
                            JSONArray jsonArray = jsonObject.getJSONArray("data");
                            ArrayList<UserData> dataList = new ArrayList<UserData>();
                            for (int i = 0; i < jsonArray.length(); i++) {
                                UserData userData = UserData.create(jsonArray.getJSONObject(i));
                                if (userData != null) {
                                    dataList.add(userData);
                                }
                            }
                            mDataList = sortUserList(dataList);
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch(Exception e) {}
                }
                callback.didReceiveData(false);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=getUser");
        httpManager.execute(Constants.ServerApiUrl, "POST", param.toString());
    }

    private ArrayList<UserData> sortUserList(ArrayList<UserData> userList) {

        ArrayList<UserData> src = new ArrayList<>(userList);
        ArrayList<UserData> dst = new ArrayList<UserData>();

        int count = src.size();
        for (int i = count - 1; i >= 0; i--) {
            int minIndex = -1;
            String minUserKana = "";

            for (int j = 0; j < src.size(); j++) {
                if (minIndex == -1) {
                    minIndex = j;
                    UserData userData = src.get(j);
                    minUserKana = userData.kanaLast + userData.kanaFirst;
                }
                UserData currentUserData = src.get(j);
                String kana = currentUserData.kanaLast + currentUserData.kanaFirst;
                if (KanaUtils.compare(kana, minUserKana) == true) {
                    minIndex = j;
                    minUserKana = kana;
                }
            }
            dst.add(src.get(minIndex));
            src.remove(minIndex);
        }
        return dst;
    }

    public interface UserRequesterCallback {
        void didReceiveData(boolean result);
    }

    public ArrayList<UserData> getUserList() {
        return mDataList;
    }

    public UserData myUserData() {

        String userId = SaveData.getInstance().userId;
        if (userId.length() == 0) return null;

        for (int i = 0; i < mDataList.size(); i++) {
            if (mDataList.get(i).userId.equals(userId)) {
                return mDataList.get(i);
            }
        }
        return null;
    }

    public UserData query(String userId) {

        for (int i = 0; i < mDataList.size(); i++) {
            if (mDataList.get(i).userId.equals(userId)) {
                return mDataList.get(i);
            }
        }
        return null;
    }

    public ArrayList<UserData> queryReservedUser(String scheduleId) {

        ArrayList<UserRequester.UserData> ret = new ArrayList<UserRequester.UserData>();
        for (int i = 0; i < mDataList.size(); i++) {
            if (mDataList.get(i).reserves.contains(scheduleId)) {
                ret.add(mDataList.get(i));
            }
        }
        return ret;
    }
}
